### DISCUSSION TITLE
Custom input adapter

### DISCUSSION THREAD

**Discussion by asraa** [nlohmann/json#2522] [Q&A]
Hi!

The open source project I work on is working on migrating away from RapidJSON to this nlohmann/json. To get function parity, I need to be a custom input adapter to track line numbers by counting newline characters. (In case there is an error in parsing, we can print out "error at line number blah" on the JSON string, and the custom object build on SAX parsing has references to line numbers).

However, I'm struggling to see how I can implement a custom input adapter to pass into `sax_parse`, without running into a failure that there is no matching constructor for `input_adapter` here https://github.com/nlohmann/json/blob/d8d8cbf6e085248e970fef2f3a047d27b195d64e/include/nlohmann/json.hpp#L6813.

Ideally, I would like to subclass `iterator_input_adapter`, override `get_character()` to also track how many newline characters are seen, and pass this custom input adapter to `sax_parse`.

I have read through https://github.com/nlohmann/json/issues/834, and don't see a resolution here. Maybe there is a better way to achieve what I need?

Thank you!

**Comment by nlohmann** [comment]
The library's detail::input_adapter(...) factory overloads only accept a fixed set of source types (iterator pairs, containers, FILE*, std::istream, contiguous byte pointers/arrays) — there's no supported extension point for subclassing iterator_input_adapter and plugging a custom instance into sax_parse/parse; input_adapter() picks the adapter type internally via overload resolution, so a hand-rolled subclass won't be selected. You don't need that hook, though: since input_adapter(first, last) works with any conforming iterator pair, you can write a thin custom iterator wrapping e.g. std::string::const_iterator whose operator++/operator* increments a shared line counter whenever a '\n' is crossed, and pass a pair of those custom iterators directly to json::sax_parse(begin, end, &handler) (or json::parse). That gives you line-number tracking during SAX parsing without touching any library internals.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
